extensions/amp-font/0.1/amp-font.js

```diff
@@ -0,0 +1,316 @@
+const DEFAULT_TIMEOUT_ = 5000;
+
+/** @private @const {string} */
+const DEFAULT_WEIGHT_ = '400';
+
+/** @private @const {string} */
+const DEFAULT_VARIANT_ = 'normal';
+
+/** @private @const {string} */
+const DEFAULT_STYLE_ = 'normal';
+
+/** @private @const {string} */
+const DEFAULT_SIZE_ = 'medium';
+
+/** @private @const {function(function())}*/
+const RAF_ =
+    requestAnimationFrame ||
+    webkitRequestAnimationFrame ||
+    mozRequestAnimationFrame ||
+    msRequestAnimationFrame ||
+    function(fn) {
+      setTimeout(fn, 10);
+    };
+
+class AmpFont extends AMP.BaseElement {
+
+
+  /** @override */
+  prerenderAllowed() {
+    return true;
+  }
+
+
+  /** @override */
+  firstAttachedCallback() {
+    this.customFontFamily_ = this.element.getAttribute('font-family');
+    this.normalizedFontFamilyName_ =
+        normalizeFontFamily_(this.customFontFamily_);
+    this.customFontWeight_ =
+        this.element.getAttribute('font-weight') || DEFAULT_WEIGHT_;
+    this.customFontStyle_ =
+        this.element.getAttribute('font-style') || DEFAULT_STYLE_;
+    this.customFontVariant_ =
+        this.element.getAttribute('font-variant') || DEFAULT_VARIANT_;
+    this.customFontSize_ =
+        this.element.getAttribute('font-size') || DEFAULT_SIZE_;
+    this.measurements_ = Object.create(null);
+    this.timeoutHandle_ = null;
+    this.hasTimedOut_ = false;
+    this.win_ = this.getWin();
+    this.document_ = this.win_.document;
+    this.documentElement_ = this.document_.documentElement;
+  }
+
+
+  /** @override */
+  buildCallback() {
+    this.startLoad_();
+  }
+
+
+  /**
+   * Start to download the font if it is already not available.
+   * @private
+   */
+  startLoad_() {
+    var addClassName = this.element.getAttribute('while-loading-class') || '';
+    this.documentElement_.classList.add(addClassName);
+    this.beginTimeout_();
+    if ('fonts' in this.document_) {
+      this.loadFontWithNativeApis_();
+    } else {
+      this.onBeforeForceLoadFont_();
+      this.forceLoadFont_();
+    }
+  }
+
+
+  /**
+   * If the browser supports font loading events use them to load the font.
+   * @private
+   */
+  loadFontWithNativeApis_() {
+    /* style | variant | weight | size/line-height | family */
+    /* font: italic small-caps bolder 16px/3 cursive; */
+    let fontString = (
+      this.customFontStyle_ + ' ' +
+      this.customFontVariant_ + ' ' +
+      this.customFontWeight_ + ' ' +
+      this.customFontSize_ + ' ' +
+      this.customFontFamily_);
+    if (this.document_.fonts.check(fontString)) {
+      this.onFontLoadSuccess_();
+    } else {
+      this.document_.fonts.load(fontString).then(() => {
+        this.onFontLoadSuccess_();
+      });
+    }
+  }
+
+
+  /**
+   * Step 1 for loading font on browsers that don't support font loading events.
+   * Creates divs hidden from the viewport and measures dimensions for default
+   * fonts.
+   * @private
+   */
+  onBeforeForceLoadFont_() {
```

Comment:
Somehow we need to do this in vsync. It's very undesirable to do sync reads during `buildCallback` pipeline.